### PR TITLE
fix: allow rerender in SlotMixin

### PR DESCRIPTION
**Pull request by tlouisse**
## What I did

Allow `SlotRerenderObject`, so that:
- complex slots can be rerendered
- scoped registries are taken care of by SlotMixin
- a wrapped renderTarget is maintained for predictable render results w/o focus and other problems
- property effects rerender the template for the particular slot that got `SlotRerenderObject` provided

```ts
type SlotRerenderObject = {
  template: TemplateResult;
  /* Add logic that will be performed after the render */
  afterRender?: Function;
};
```

### CODE REVIEW

File: packages/ui/components/core/test/SlotMixin.test.js [hunk 1]
```diff
@@ -195,102 +142,361 @@ describe('SlotMixin', () => {
...
+          }
+        },
+      );
+      const el = /** @type {* & SlotHost} */ (await fixture(`<${tag}></${tag}>`));
+
+      el._focusableNode.focus();
+      expect(document.activeElement).to.equal(el._focusableNode);
+
+      // el.currentValue = 1;
+      // await el.updateComplete;
```
Comment (daKmoR) [inline packages/ui/components/core/test/SlotMixin.test.js:228]: this is not executed? so what is it testing then? 🤔

Comment (tlouisse) [reply packages/ui/components/core/test/SlotMixin.test.js:228]: Valid point. These are some leftovers of a previous approach. Will look into it

Comment (tlouisse) [reply packages/ui/components/core/test/SlotMixin.test.js]: Should be fixed 👍

File: docs/fundamentals/systems/core/SlotMixin.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+`<my-input label="My label"></my-input>`
+
+- A more advanced label can be provided like this:
+  `<my-input> <label slot="label"><my-icon aria-hidden="true"></my-icon>My label</label> </my-input>`
+- In the property/attribute case, SlotMixin adds the `<label slot="label">` under the hood. **unless** the developer already provided the slot.
+  This will make sure that the slot provided by the user always takes precedence and only one slot instance will be available in light dom per slot.
+
+### Default slot
+
+As can be seen in the example above, '' can be used to add content to the default slot
```
Comment (gerjanvangeest) [inline docs/fundamentals/systems/core/SlotMixin.md:64]:
```suggestion
As can be seen in the example below, '' can be used to add content to the default slot
```

File: packages/ui/components/core/src/SlotMixin.js [hunk 3]
```diff
@@ -6,10 +6,145 @@ import { isTemplateResult } from 'lit/directive-helpers.js';
...
+ * @typedef {import('../types/SlotMixinTypes.js').SlotFunctionResult} SlotFunctionResult
+ * @typedef {import('../types/SlotMixinTypes.js').SlotRerenderObject} SlotRerenderObject
  * @typedef {import('lit').LitElement} LitElement
  */
 
+const isRerenderConfig = (/** @type {SlotFunctionResult} */ o) =>
+  !Array.isArray(o) && typeof o === 'object' && 'template' in o;
+
 /**
+ * # Systems >> Core >> SlotMixin ||20
```
Comment (gerjanvangeest) [inline packages/ui/components/core/src/SlotMixin.js:18]: Doesn't make sense to have the documentation double, keep 1 source of truth.

File: packages/ui/components/input-tel-dropdown/src/LionInputTelDropdown.js [hunk 4]
```diff
@@ -259,9 +268,9 @@ export class LionInputTelDropdown extends LionInputTel {
   updated(changedProperties) {
     super.updated(changedProperties);
 
-    if (changedProperties.has('_needsLightDomRender')) {
-      this.__renderDropdown();
-    }
+    // if (changedProperties.has('_needsLightDomRender')) {
+    //   this.__renderDropdown();
+    // }
```
Comment (gerjanvangeest) [inline packages/ui/components/input-tel-dropdown/src/LionInputTelDropdown.js:273]:
```suggestion

```

File: packages/ui/components/input-tel-dropdown/test-suites/LionInputTelDropdown.suite.js [hunk 5]
```diff
@@ -92,7 +93,7 @@ export function runInputTelDropdownSuite({ klass } = { klass: LionInputTelDropdo
     it('syncs value of dropdown on init if input has no value does not influence interaction states', async () => {
       const el = await fixture(html` <${tag}></${tag}> `);
       // TODO find out why its get dirty again
```
Comment (gerjanvangeest) [inline packages/ui/components/input-tel-dropdown/test-suites/LionInputTelDropdown.suite.js:95]:
```suggestion
```

File: packages/ui/components/input-tel-dropdown/test-suites/LionInputTelDropdown.suite.js [hunk 6]
```diff
@@ -71,6 +71,7 @@ export function runInputTelDropdownSuite({ klass } = { klass: LionInputTelDropdo
 
     it('syncs value of dropdown on init if input has no value', async () => {
       const el = await fixture(html` <${tag}></${tag}> `);
+      // await el.updateComplete;
```
Comment (gerjanvangeest) [inline packages/ui/components/input-tel-dropdown/test-suites/LionInputTelDropdown.suite.js:74]:
```suggestion
```

File: .changeset/warm-forks-press.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+'@lion/ui': patch
+---
+
+Align with SlotMixin rerender functionality + fix interaction state synchronization
```
Comment (gerjanvangeest) [inline .changeset/warm-forks-press.md:5]:
```suggestion
[input-tel-dropdown] Align with SlotMixin rerender functionality + fix interaction state synchronization
```

File: docs/fundamentals/systems/core/SlotMixin.md (hunk 2 above)
Comment (tlouisse) [reply docs/fundamentals/systems/core/SlotMixin.md:64]: 👍

File: packages/ui/components/core/src/SlotMixin.js (hunk 3 above)
Comment (tlouisse) [reply packages/ui/components/core/src/SlotMixin.js:18]: Yeah agree with SSOT. Also like the principle of having docs as close to code as possible (will be better maintained and read by devs using code base), but we don't have a build process that allows for this (like angularjs had back in the days).
Compromise: I add a short summary and mention the existence of the extended docs

File: packages/ui/components/input-tel-dropdown/src/LionInputTelDropdown.js (hunk 4 above)
Comment (tlouisse) [reply packages/ui/components/input-tel-dropdown/src/LionInputTelDropdown.js:273]: 👍

File: packages/ui/components/input-tel-dropdown/test-suites/LionInputTelDropdown.suite.js (hunk 5 above)
Comment (tlouisse) [reply packages/ui/components/input-tel-dropdown/test-suites/LionInputTelDropdown.suite.js:95]: 👍

File: .changeset/warm-forks-press.md (hunk 7 above)
Comment (tlouisse) [reply .changeset/warm-forks-press.md:5]: 👍

**Review by gerjanvangeest — APPROVED** [review]

**Pull request merged**